File: src/data/achievements.ts
import { AchievementDef, AchievementContext, PlayerProfileData, SessionSummary } from '../types';

const categoryCorrect = (ctx: AchievementContext, category: string): number =>
  ctx.categoryCorrect[category] ?? 0;

export const ACHIEVEMENTS: AchievementDef[] = [
  {
    id: 'first_step',
    name: 'First Step',
    description: 'Answer your first question',
    icon: 'foot-print',
    condition: (c) => c.totalAnswered >= 1,
  },
  {
    id: 'streak_3',
    name: 'Hat-Trick',
    description: 'Get a 3 answer streak',
    icon: 'fire',
    condition: (c) => c.bestStreak >= 3,
  },
  {
    id: 'streak_5',
    name: 'Hot Streak',
    description: 'Get a 5 answer streak',
    icon: 'fire-circle',
    condition: (c) => c.bestStreak >= 5,
  },
  {
    id: 'streak_10',
    name: 'Trial by Fire',
    description: 'Get a 10 answer streak',
    icon: 'lightning-bolt',
    condition: (c) => c.bestStreak >= 10,
  },
  {
    id: 'perfect_game',
    name: 'Flawless Victory',
    description: 'Finish a game with every answer correct',
    icon: 'trophy',
    condition: (c) => c.perfectGames >= 1 || c.sessionPerfect,
  },
  {
    id: 'no_lifeline',
    name: 'Apne Dum Par',
    description: 'Score 8+ in a game without using lifelines',
    icon: 'arm-flex',
    condition: (c) => c.sessionLifelinesUsed === 0 && c.sessionCorrect >= 8,
  },
  {
    id: 'historian',
    name: 'History Buff',
    description: 'Answer 10 History questions correctly',
    icon: 'bank',
    condition: (c) => categoryCorrect(c, 'history') >= 10,
  },
  {
    id: 'culture_fan',
    name: 'Culture Master',
    description: 'Answer 10 Culture questions correctly',
    icon: 'theater',
    condition: (c) => categoryCorrect(c, 'culture') >= 10,
  },
  {
    id: 'festival_fan',
    name: 'Festival Fan',
    description: 'Answer 10 Festival questions correctly',
    icon: 'party-popper',
    condition: (c) => categoryCorrect(c, 'festivals') >= 10,
  },
  {
    id: 'myth_scholar',
    name: 'Mythology Scholar',
    description: 'Answer 10 Mythology questions correctly',
    icon: 'book-open-page-variant',
    condition: (c) => categoryCorrect(c, 'mythology') >= 10,
  },
  {
    id: 'geographer',
    name: 'Geography Whiz',
    description: 'Answer 10 Geography questions correctly',
    icon: 'earth',
    condition: (c) => categoryCorrect(c, 'geography') >= 10,
  },
  {
    id: 'level_5',
    name: 'Scholar',
    description: 'Reach level 5',
    icon: 'shield-star',
    condition: (c) => c.level >= 5,
  },
  {
    id: 'level_10',
    name: 'Grandmaster',
    description: 'Reach level 10',
    icon: 'crown',
    condition: (c) => c.level >= 10,
  },
  {
    id: 'correct_50',
    name: 'Half-Century',
    description: '50 total correct answers',
    icon: 'star-circle',
    condition: (c) => c.totalCorrect >= 50,
  },
  {
    id: 'correct_100',
    name: 'Centurion',
    description: '100 total correct answers',
    icon: 'star-shooting',
    condition: (c) => c.totalCorrect >= 100,
  },
  {
    id: 'games_10',
    name: 'Regular Player',
    description: 'Play 10 games',
    icon: 'gamepad-variant',
    condition: (c) => c.gamesPlayed >= 10,
  },
];

const byId = new Map<string, AchievementDef>(ACHIEVEMENTS.map((a) => [a.id, a]));

export function getAchievement(id: string): AchievementDef {
  const a = byId.get(id);
  if (!a) throw new Error(`Unknown achievement: ${id}`);
  return a;
}

export function checkNewAchievements(
  profile: PlayerProfileData,
  sessionSummary: SessionSummary
): AchievementDef[] {
  const ctx: AchievementContext = {
    totalXp: profile.totalXp,
    totalAnswered: profile.totalAnswered,
    totalCorrect: profile.totalCorrect,
    bestStreak: profile.bestStreak,
    gamesPlayed: profile.gamesPlayed,
    perfectGames: profile.perfectGames,
    level: levelForXp(profile.totalXp),
    categoryCorrect: { ...profile.categoryCorrect },
    sessionPerfect: sessionSummary.perfect,
    sessionBestStreak: sessionSummary.bestStreak,
    sessionXp: sessionSummary.xpEarned,
    sessionCorrect: sessionSummary.correct,
    sessionLifelinesUsed: sessionSummary.lifelinesUsed,
  };

  const earned = new Set(profile.achievements);
  return ACHIEVEMENTS.filter((a) => !earned.has(a.id) && a.condition(ctx));
}

export function xpToReach(level: number): number {
  return 50 * (level - 1) * level;
}

export function levelForXp(totalXp: number): number {
  let level = 1;
  while (xpToReach(level + 1) <= totalXp) {
    level += 1;
  }
  return level;
}

const LEVEL_TITLES: [number, string][] = [
  [13, 'Grandmaster'],
  [11, 'Illuminated'],
  [9, 'Mentor'],
  [7, 'Pundit'],
  [5, 'Scholar'],
  [3, 'Apprentice'],
  [1, 'Novice'],
];

export function titleForLevel(level: number): string {
  for (const [minLevel, title] of LEVEL_TITLES) {
    if (level >= minLevel) return title;
  }
  return LEVEL_TITLES[LEVEL_TITLES.length - 1][1];
}

File: gyaanquest/core/achievements.py
"""Achievements — badges unlocked by player feats.

Each achievement has a condition over a context object exposing:
    profile fields: total_xp, total_answered, total_correct, best_streak,
                    games_played, perfect_games, level,
                    category_correct (dict)
    session fields: session_perfect, session_best_streak, session_xp,
                    session_correct, session_lifelines_used
"""
from __future__ import annotations

from dataclasses import dataclass
from types import SimpleNamespace
from typing import Callable


@dataclass(frozen=True)
class Achievement:
    id: str
    name: str
    description: str
    icon: str
    condition: Callable[[SimpleNamespace], bool]


def _cat(ctx, category):
    return ctx.category_correct.get(category, 0)


ACHIEVEMENTS = [
    Achievement("first_step", "First Step", "Answer your first question",
                "foot-print", lambda c: c.total_answered >= 1),
    Achievement("streak_3", "Hat-Trick", "Get a 3 answer streak",
                "fire", lambda c: c.best_streak >= 3),
    Achievement("streak_5", "Hot Streak", "Get a 5 answer streak",
                "fire-circle", lambda c: c.best_streak >= 5),
    Achievement("streak_10", "Trial by Fire", "Get a 10 answer streak",
                "lightning-bolt", lambda c: c.best_streak >= 10),
    Achievement("perfect_game", "Flawless Victory", "Finish a game with every answer correct",
                "trophy", lambda c: c.perfect_games >= 1 or c.session_perfect),
    Achievement("no_lifeline", "Apne Dum Par", "Score 8+ in a game without using lifelines",
                "arm-flex",
                lambda c: c.session_lifelines_used == 0 and c.session_correct >= 8),
    Achievement("historian", "History Buff", "Answer 10 History questions correctly",
                "bank", lambda c: _cat(c, "history") >= 10),
    Achievement("culture_fan", "Culture Master", "Answer 10 Culture questions correctly",
                "theater", lambda c: _cat(c, "culture") >= 10),
    Achievement("festival_fan", "Festival Fan", "Answer 10 Festival questions correctly",
                "party-popper", lambda c: _cat(c, "festivals") >= 10),
    Achievement("myth_scholar", "Mythology Scholar", "Answer 10 Mythology questions correctly",
                "book-open-page-variant", lambda c: _cat(c, "mythology") >= 10),
    Achievement("geographer", "Geography Whiz", "Answer 10 Geography questions correctly",
                "earth", lambda c: _cat(c, "geography") >= 10),
    Achievement("level_5", "Scholar", "Reach level 5",
                "shield-star", lambda c: c.level >= 5),
    Achievement("level_10", "Grandmaster", "Reach level 10",
                "crown", lambda c: c.level >= 10),
    Achievement("correct_50", "Half-Century", "50 total correct answers",
                "star-circle", lambda c: c.total_correct >= 50),
    Achievement("correct_100", "Centurion", "100 total correct answers",
                "star-shooting", lambda c: c.total_correct >= 100),
    Achievement("games_10", "Regular Player", "Play 10 games",
                "gamepad-variant", lambda c: c.games_played >= 10),
]

_BY_ID = {a.id: a for a in ACHIEVEMENTS}


def get(achievement_id: str) -> Achievement:
    return _BY_ID[achievement_id]


def check_new(profile, session_summary) -> list[Achievement]:
    """Return achievements newly earned (not already in profile.achievements)."""
    ctx = SimpleNamespace(
        total_xp=profile.total_xp,
        total_answered=profile.total_answered,
        total_correct=profile.total_correct,
        best_streak=profile.best_streak,
        games_played=profile.games_played,
        perfect_games=profile.perfect_games,
        level=profile.level,
        category_correct=dict(profile.category_correct),
        session_perfect=session_summary.get("perfect", False),
        session_best_streak=session_summary.get("best_streak", 0),
        session_xp=session_summary.get("xp_earned", 0),
        session_correct=session_summary.get("correct", 0),
        session_lifelines_used=session_summary.get("lifelines_used", 0),
    )
    earned = set(profile.achievements)
    return [a for a in ACHIEVEMENTS if a.id not in earned and a.condition(ctx)]
